bundles/org.openhab.binding.openthermgateway/src/main/java/org/openhab/binding/openthermgateway/handler/OpenThermGatewayHandler.java

```diff
@@ -108,184 +94,177 @@ public void handleCommand(ChannelUID channelUID, Command command) {
-        if (!explicitDisconnect && conf != null && conf.connectionRetryInterval > 0) {
-            logger.debug("Scheduling to reconnect in {} seconds.", conf.connectionRetryInterval);
-            reconnectTask = scheduler.schedule(this::connect, conf.connectionRetryInterval, TimeUnit.SECONDS);
+        if (conn != null && conn.isConnected()) {
+            conn.sendCommand(gatewayCommand);
+        } else {
+            logger.debug("Unable to send command {}: connector not connected", gatewayCommand.toFullString());
         }
     }
 
-    @Override
     public void receiveMessage(Message message) {
-        if (DataItemGroup.dataItemGroups.containsKey(message.getID())) {
-            DataItem[] dataItems = DataItemGroup.dataItemGroups.get(message.getID());
+        int msgId = message.getID();
 
-            for (DataItem dataItem : dataItems) {
-                String channelId = dataItem.getSubject();
+        if (!DataItemGroup.DATAITEMGROUPS.containsKey(msgId)) {
+            logger.debug("Unsupported message id {}", msgId);
+            return;
+        }
 
-                if (!OpenThermGatewayBindingConstants.SUPPORTED_CHANNEL_IDS.contains(channelId)
-                        || (dataItem.getFilteredCode() != null && dataItem.getFilteredCode() != message.getCode())) {
-                    continue;
-                }
+        for (Thing thing : getThing().getThings()) {
+            BaseDeviceHandler handler = (BaseDeviceHandler) thing.getHandler();
 
-                State state = null;
-
-                switch (dataItem.getDataType()) {
-                    case FLAGS:
-                        state = OnOffType.from(message.getBit(dataItem.getByteType(), dataItem.getBitPos()));
-                        break;
-                    case UINT8:
-                    case UINT16:
-                        state = new DecimalType(message.getUInt(dataItem.getByteType()));
-                        break;
-                    case INT8:
-                    case INT16:
-                        state = new DecimalType(message.getInt(dataItem.getByteType()));
-                        break;
-                    case FLOAT:
-                        float value = message.getFloat();
-                        @Nullable
-                        Unit<?> unit = dataItem.getUnit();
-                        state = (unit == null) ? new DecimalType(value) : new QuantityType<>(value, unit);
-                        break;
-                    case DOWTOD:
-                        break;
-                }
+            if (handler != null) {
+                handler.receiveMessage(message);
+            }
+        }
+    }
 
-                if (state != null) {
-                    logger.debug("Received update for channel '{}': {}", channelId, state);
-                    updateState(channelId, state);
-                }
+    @Override
+    public void connectionStateChanged(ConnectionState state) {
```

Comment:
The socket connecter is running in its own thread, with is a different thread than the OH framework, and in my tests I found that sometimes if the OH framework would block or is delayed, that causes the socket connector's `run()` method to be delayed (which can lead to e.g. connection errors or timeouts. Instead I suggest to fully decouple the two threads by interposing a separate scheduler task as follows...

```
 @Override
    public void connectionStateChanged(ConnectionState state) {
        scheduler.submit(() -> connectionStateChangedTask());
    }
    
    private void connectionStateChangedTask(ConnectionState state) {
```